### ISSUE TITLE
Consistency check failed: AV in clr at this callstack

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/coreclr#23317]
Compile following program, run crossgen on it, and run it with corerun.

```csharp
delegate int Del(object p);

abstract class Base
{
    public abstract int Target<U>(object p);
}

class Middle : Base
{
    public override int Target<U>(object p) => 123;
}

class Top : Middle
{
    public override int Target<U>(object p) => 456;

    public Del TestA<U>()
    {
        return new Del(base.Target<U>);
    }

    public Del TestB<U>()
    {
        return new Del(Target<U>);
    }

    public static int Main()
    {
        return (new Top().TestA<object>()(null)) + (new Top().TestB<object>()(null));
    }
}
```

Observed:

```
Assert failure(PID 33796 [0x00008404], Thread: 32104 [0x7d68]): Consistency check failed: AV in clr at this callstack:
------
CORECLR! tagCOR_ILMETHOD_TINY::IsTiny + 0x16 (0x00007ffb`79f79d56)
CORECLR! DecoderInit + 0x35 (0x00007ffb`79f7b085)
CORECLR! `COR_ILMETHOD_DECODER::COR_ILMETHOD_DECODER'::`5'::__Body::Run + 0x58 (0x00007ffb`7ac93278)
CORECLR! COR_ILMETHOD_DECODER::COR_ILMETHOD_DECODER + 0x71 (0x00007ffb`7ac92f51)
CORECLR! ETW::MethodLog::SendMethodJitStartEvent + 0x2FF (0x00007ffb`7a133a0f)
CORECLR! ETW::MethodLog::MethodJitting + 0x300 (0x00007ffb`7a12aaf0)
CORECLR! MethodDesc::JitCompileCodeLockedEventWrapper + 0x3F5 (0x00007ffb`7a3200b5)
CORECLR! MethodDesc::JitCompileCode + 0x52B (0x00007ffb`7a31f70b)
CORECLR! MethodDesc::PrepareILBasedCode + 0x298 (0x00007ffb`7a321c58)
CORECLR! MethodDesc::PrepareCode + 0x1D9 (0x00007ffb`7a3216b9)
CORECLR! MethodDesc::PrepareInitialCode + 0x1C7 (0x00007ffb`7a321e87)
CORECLR! MethodDesc::DoPrestub + 0xA52 (0x00007ffb`7a319442)
CORECLR! PreStubWorker + 0x45A (0x00007ffb`7a325caa)
CORECLR! ThePreStub + 0x55 (0x00007ffb`7a606ce5)
PRE.NI! <no symbol> + 0x0 (0x00007ffb`db692c10)
CORECLR! CallDescrWorkerInternal + 0x83 (0x00007ffb`7a85e413)
CORECLR! CallDescrWorkerWithHandler + 0x1C3 (0x00007ffb`7a4cf6e3)
CORECLR! MethodDescCallSite::CallTargetWorker + 0xB02 (0x00007ffb`7a4d01f2)
CORECLR! MethodDescCallSite::Call_RetArgSlot + 0xF4 (0x00007ffb`7a0c3954)
```

Expected: program succeeds and returns 579.

Hit while trying to understand the test failures that are the reason for #23291. This is a pre-existing bug in ReadyToRun that will not be fixed by rolling back my change. We can likely re-instate my change after fixing this (#23291 is going to fix the existing test failures because the cargo cults I deleted were papering over this ready to run bug for interfaces, but R2R virtual delegates have a more fundamental problem somewhere).

**Comment by MichalStrehovsky** [comment]
@sergiy-k do you know who owns R2R in CoreCLR these days? When this is hit, it's long after we did the wrong thing and I couldn't figure out where we took the wrong turn in the time I allotted myself for it.

**Comment by sergiy-k** [comment]
@fadimounir, could you please take a look?

**Comment by fadimounir** [comment]
I'll take a look

**Comment by fadimounir** [comment]
Fix: #24383
